http/middleware.ts

```diff
@@ -0,0 +1,54 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+
+/**
+ * Middleware function based on {@linkcode Deno.ServeHandlerInfo} but an added
+ * `next()` function which calls the next middleware function in the middleware
+ * chain.
+ */
+export type MiddlewareFunction = (
+  request: Request,
+  info: Deno.ServeHandlerInfo,
+  next: () => Response | Promise<Response>,
+) => Response | Promise<Response>;
+
+/**
+ * Creates a {@linkcode Deno.ServeHandler} from the given middleware chain,
+ * which can then be passed to {@linkcode Deno.serve}.
+ *
+ * @param middlewares Middleware chain
+ *
+ * @example
+ * ```ts
+ * import {
+ *   type MiddlewareFunction,
+ *   createHandler,
+ * } from "https://deno.land/std@$STD_VERSION/http/middleware.ts";
+ *
+ * const middleware1: MiddlewareFunction = async (_request, _info, next) => {
+ *   const start = performance.now();
+ *   const response = await next();
+ *   const duration = performance.now() - start;
+ *   response.headers.set("x-request-time", duration.toString());
+ *   return response;
+ * };
+ *
+ * const middleware2: MiddlewareFunction = (request, info) => {
+ *   return Response.json({ request, info });
+ * };
+ *
+ * const handler = createHandler([])
+ * ```
+ */
+export function createHandler(
+  middlewares: MiddlewareFunction[],
+): Deno.ServeHandler {
+  return (request, info) => {
+    function compose(index: number): Response | Promise<Response> {
+      if (index >= middlewares.length) {
+        throw new RangeError("Middleware chain exhausted");
```

Comment:
Does this error make sense?

Reply:
I'm not sure if a `RangeError` is the right Error class for this. Should this be a custom error?

Reply:
I'm not certain either. I'd be keen to hear other's thoughts.

Reply:
The way the `createHandler` function is designed here assumes that the last middleware never calls `next()`. So in that sense it's fine, although it prevents composing `createHandler` itself. Not sure if that is a goal here though.

```js
const a = createHandler([m1, m2, m3]);
const b = createHandler([m1, m2, m3]);

// Not possible
const c = createHandler([a, b]);
```

Reply:
Yeah, it wasn't one of my goals to do that... I don't know the advantage of doing it and whether there'd be sufficient use cases for that.